fetch_index.py

```diff
@@ -130,237 +218,320 @@ def index_lists(self, sites, collection, index):
+                                        value[1]),
+                                    "Error while indexing the listitem for list: %s to the workplace." % (
+                                        value[1])
+                                    )
+
             responses.append(document)
         return responses
 
-    def index_permissions(
-        self,
-        key,
-        collection,
-        site=None,
-        list_name=None,
-        list_url=None,
-        itemid=None,
-    ):
-        unique = False
+    def get_roles(self, key, site, list_url, list_name, itemid):
+        """ Checks the permissions and returns the user roles.
+            :param key: key, a string value
+            :param site: site name to check the permission
+            :param list_url: list url to access the list
+            :param list_name: list name to check the permission
+            :param itemid: item id to check the permission
+            Returns:
+                roles: user roles
+        """
         if key == SITES:
             rel_url = urljoin(self.sharepoint_host, site)
-            unique = self.permissions.check_permissions(key, rel_url)
-            if unique:
-                roles = self.permissions.fetch_users(key, rel_url)
+            roles = self.permissions.fetch_users(key, rel_url)
+
         elif key == LISTS:
             rel_url = urljoin(self.sharepoint_host, list_url)
-            unique = self.permissions.check_permissions(
+            roles = self.permissions.fetch_users(
                 key, rel_url, title=list_name
             )
-            if unique:
-                roles = self.permissions.fetch_users(
-                    key, rel_url, title=list_name
-                )
 
         elif key == ITEMS:
-
             rel_url = urljoin(self.sharepoint_host, list_url)
-            unique = self.permissions.check_permissions(
+            roles = self.permissions.fetch_users(
                 key, rel_url, title=list_name, id=itemid
             )
-            if unique:
-                roles = self.permissions.fetch_users(
-                    key, rel_url, title=list_name, id=itemid
-                )
 
-        if not unique:
-            host_url = urljoin(
-                self.sharepoint_host, f"/sites/{collection}/_api/"
+        return roles, rel_url
+
+    def workplace_add_permission(self, user_name, permission):
+        """This method when invoked would index the permission provided in the paramater
+            for the user in paramter user_name
+            :param user_name: a string value denoting the username of the user
+            :param permission: permission that needs to be provided to the user
+        """
+        try:
+            self.ws_client.add_user_permissions(
+                content_source_id=self.ws_source,
+                http_auth=self.ws_token,
+                user=user_name,
+                body={
+                    "permissions": [permission]
```

Comment:
Hey, I guess Enterprise Search client method `add_user_permissions` requires that the value passed to the `permissions` field must be an array. We are adding one permission at a time so I had to pass the `permission` variable inside an array.

Reply:
[Here](https://github.com/elastic/enterprise-search-python/blob/1788413218badc01e2da23ac290698de40117f8c/elastic_enterprise_search/client/_workplace_search.py#L1416) is the client code, which links to [the documentation of the API](https://www.elastic.co/guide/en/workplace-search/master/workplace-search-document-permissions-api.html#add-one), where it states 
>Add one or more permission for a given user.

So I think it will be better to make this method more general to accept an array of permissions, and to index permissions in batch if possible.

Reply:
I got your point. You're absolutely right @wangch079 . We should index permissions in batches as it would reduce the number of API calls. I will look into the documentation once again and come up with the required changes

Reply:
Done. Commit link: [e641703](https://github.com/elastic/workplace-search-sharepoint16-connector/pull/2/commits/e641703806f7fca06655dfcfccb9e927f15afc6f)